### PR TITLE
When a copy task filters lines via closure, remove line if null is returned

### PR DESCRIPTION
**Pull request by benblank**
This feature was [requested on the forum last year](https://www.google.com/url?q=https%3A%2F%2Fdiscuss.gradle.org%2Ft%2Fit-should-be-possible-to-remove-lines-using-filter-closure%2F518%2F4&sa=D&sntz=1&usg=AFQjCNEILOx0v_mkoNVqMnoCtIMggsfIDw), but never implemented.  There is [a gradle-dev thread](https://groups.google.com/forum/#!topic/gradle-dev/miICu-menSk) for the proposed change.

This change required a minor restructuring of LineFilter.  Rather than tracking its state implicitly (with a null `transformedLine` representing end-of-file, except at beginning-of-file), it now tracks state explicitly, allowing it to loop continuously over input lines as the closure returns `null` without emitting output lines.

Unit and integration tests have been added for both the common case of only specific lines being removed in this manner as well as the edge case of *all* lines being removed.  The file `copyTestResources/src/two/two.b` was commandeered for the integration tests, as its contents were not being used in any other test and adding an entirely new file would have impacted other tests.

### CODE REVIEW

File: subprojects/core/src/main/groovy/org/gradle/api/file/ContentFilterable.java [hunk 1]
```diff
@@ -61,7 +61,7 @@
 
     /**
      * Adds a content filter based on the provided closure.  The Closure will be called with each line (stripped of line
-     * endings) and should return a String to replace the line.
+     * endings) and should return a String to replace the line or {@code null} to remove the line.
```
Comment (alkemist) [inline subprojects/core/src/main/groovy/org/gradle/api/file/ContentFilterable.java]: Can you add some text to describe what happens if the closure always returns null please.

**Comment by alkemist** [comment]
Quality contribution. Very thorough. Thanks. 

Just waiting on some more votes from other devs.

File: subprojects/core/src/main/groovy/org/gradle/api/file/ContentFilterable.java (hunk 1 above)
Comment (benblank) [inline subprojects/core/src/main/groovy/org/gradle/api/file/ContentFilterable.java]: Can do.  What's the preferred method for updating a PR — amend the original commit or add a new one?

Comment (bigdaz) [inline subprojects/core/src/main/groovy/org/gradle/api/file/ContentFilterable.java]: Adding another commit to the PR is fine.

**Comment by alkemist** [comment]
Ready to merge this.

@benblank have you signed the http://gradle.org/contributor-license-agreement/ ?

**Comment by benblank** [comment]
Indeed I have.

**Comment by alkemist** [comment]
Great, thanks.
